File: packages/frontend/src/api/order.api.ts
import api from './client';
import type { ApiResponse, Order, OrderItem } from '@/types';

export interface CreateOrderPayload {
  items: { productId: string; quantity: number; productImageId?: string }[];
  traderId?: string;
  orderMode?: 'DIRECT' | 'MANAGED';
  note?: string;
  /** Optional `YYYY-MM-DD` — need-by end of that day UTC. */
  customerNeedBy?: string;
}

export const orderApi = {
  create: (data: CreateOrderPayload) =>
    api.post<ApiResponse<Order>>('/orders', data).then((r) => r.data.data),

  list: () =>
    api.get<ApiResponse<Order[] | OrderItem[]>>('/orders').then((r) => r.data.data),

  getById: (id: string) =>
    api.get<ApiResponse<Order>>(`/orders/${id}`).then((r) => r.data.data),

  confirm: (id: string) =>
    api.post<ApiResponse<Order>>(`/orders/${id}/confirm`).then((r) => r.data.data),

  modifyItems: (id: string, items: { itemId: string; requestedQty: number }[]) =>
    api.post<ApiResponse<Order>>(`/orders/${id}/modify`, { items }).then((r) => r.data.data),

  cancel: (id: string) =>
    api.post<ApiResponse<Order>>(`/orders/${id}/cancel`).then((r) => r.data.data),

  traderAlerts: () =>
    api.get<ApiResponse<(Order & { alertType: string; pendingCount: number; rejectedCount: number })[]>>('/orders/trader/alerts').then((r) => r.data.data),

  takeControl: (id: string) =>
    api.post<ApiResponse<Order>>(`/orders/${id}/take-control`).then((r) => r.data.data),

  releaseToVendors: (id: string) =>
    api.post<ApiResponse<Order>>(`/orders/${id}/release-to-vendors`).then((r) => r.data.data),

  traderAdjust: (
    id: string,
    body: {
      items?: { itemId: string; requestedQty?: number; unitPrice?: number | null }[];
      note?: string | null;
    },
  ) => api.post<ApiResponse<Order>>(`/orders/${id}/trader-adjust`, body).then((r) => r.data.data),

  setTraderCounterPrice: (itemId: string, unitPrice: number) =>
    api.put<ApiResponse<OrderItem>>(`/orders/items/${itemId}/counter-price`, { unitPrice }).then((r) => r.data.data),
};

File: packages/mobile/lib/core/api/order_api.dart
import 'package:dio/dio.dart';

import '../network/api_response.dart';

class OrderApi {
  OrderApi(this._dio);
  final Dio _dio;

  Future<Map<String, dynamic>> create(Map<String, dynamic> data) async {
    final res = await _dio.post<Map<String, dynamic>>('/orders', data: data);
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<dynamic> list() async {
    final res = await _dio.get<Map<String, dynamic>>('/orders');
    return ApiResponse.unwrapData(res.data!, (d) => d);
  }

  Future<Map<String, dynamic>> getById(String id) async {
    final res = await _dio.get<Map<String, dynamic>>('/orders/$id');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> confirm(String id) async {
    final res = await _dio.post<Map<String, dynamic>>('/orders/$id/confirm');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> modifyItems(String id, List<Map<String, dynamic>> items) async {
    final res = await _dio.post<Map<String, dynamic>>('/orders/$id/modify', data: {'items': items});
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> cancel(String id) async {
    final res = await _dio.post<Map<String, dynamic>>('/orders/$id/cancel');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<List<dynamic>> traderAlerts() async {
    final res = await _dio.get<Map<String, dynamic>>('/orders/trader/alerts');
    return ApiResponse.unwrapData(res.data!, (d) => d as List<dynamic>);
  }

  Future<Map<String, dynamic>> takeControl(String id) async {
    final res = await _dio.post<Map<String, dynamic>>('/orders/$id/take-control');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> releaseToVendors(String id) async {
    final res = await _dio.post<Map<String, dynamic>>('/orders/$id/release-to-vendors');
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> traderAdjust(String id, Map<String, dynamic> body) async {
    final res = await _dio.post<Map<String, dynamic>>('/orders/$id/trader-adjust', data: body);
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }

  Future<Map<String, dynamic>> setTraderCounterPrice(String itemId, double unitPrice) async {
    final res = await _dio.put<Map<String, dynamic>>(
      '/orders/items/$itemId/counter-price',
      data: {'unitPrice': unitPrice},
    );
    return ApiResponse.unwrapData(res.data!, (d) => d as Map<String, dynamic>);
  }
}
